src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsa.cs

```diff
@@ -90,21 +99,355 @@ public virtual byte[] SignData(byte[] data, int offset, int count, HashAlgorithm
+            int offset,
+            int count,
+            HashAlgorithmName hashAlgorithm,
+            DSASignatureFormat signatureFormat)
         {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+            if (offset < 0 || offset > data.Length)
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            if (count < 0 || count > data.Length - offset)
+                throw new ArgumentOutOfRangeException(nameof(count));
             if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
+            if (!signatureFormat.IsKnownValue())
+                throw DSASignatureFormatHelpers.CreateUnknownValueException(signatureFormat);
+
+            return SignDataCore(new ReadOnlySpan<byte>(data, offset, count), hashAlgorithm, signatureFormat);
+        }
+
+        /// <summary>
+        ///   Computes the hash value of the specified data and signs it using the specified signature format.
+        /// </summary>
+        /// <param name="data">The data to sign.</param>
+        /// <param name="hashAlgorithm">The hash algorithm to use to create the hash value.</param>
+        /// <param name="signatureFormat">The encoding format to use for the signature.</param>
+        /// <returns>
+        ///   The ECDSA signature for the specified data.
+        /// </returns>
+        /// <exception cref="CryptographicException">
+        ///   An error occurred in the hashing or signing operation.
+        /// </exception>
+        protected virtual byte[] SignDataCore(
+            ReadOnlySpan<byte> data,
+            HashAlgorithmName hashAlgorithm,
+            DSASignatureFormat signatureFormat)
+        {
+            Span<byte> signature = stackalloc byte[SignatureStackBufSize];
+            int maxSignatureSize = GetMaxSignatureSize(signatureFormat);
+            byte[]? rented = null;
+            int bytesWritten = 0;
+
+            if (maxSignatureSize > signature.Length)
+            {
+                // Use the shared pool because the buffer is passed out.
+                rented = ArrayPool<byte>.Shared.Rent(maxSignatureSize);
+                signature = rented;
+            }
+
+            try
             {
+                if (!TrySignDataCore(data, signature, hashAlgorithm, signatureFormat, out bytesWritten))
+                {
+                    Debug.Fail($"GetMaxSignatureSize returned insufficient size for format {signatureFormat}");
+                    throw new CryptographicException();
+                }
+
+                return signature.Slice(0, bytesWritten).ToArray();
+            }
+            finally
+            {
+                if (rented != null)
+                {
+                    CryptographicOperations.ZeroMemory(rented.AsSpan(0, bytesWritten));
+                    ArrayPool<byte>.Shared.Return(rented);
```

Comment:
Same comment about `ArrayPool` and _finally_ blocks here. (This comment applies elsewhere throughout these files.)